warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfo.java

```diff
@@ -143,75 +137,44 @@ public IndexInfo union(IndexInfo o) {
         
         if (null != first.myNode && first.myNode != o.myNode) {
             
             JexlNode sourceNode = getSourceNode(first.myNode);
             JexlNode topLevelOr = getOrNode(sourceNode);
             
             if (null == topLevelOr) {
-                topLevelOr = sourceNode;
-                // add the source node
-                nodesMap.put(nodeToKey(sourceNode), first.myNode);
+                nodeSet.add(sourceNode);
             } else {
                 for (int i = 0; i < topLevelOr.jjtGetNumChildren(); i++) {
-                    JexlNode baseNode = getSourceNode(topLevelOr.jjtGetChild(i));
-                    nodesMap.put(nodeToKey(baseNode), topLevelOr.jjtGetChild(i));
+                    nodeSet.add(topLevelOr.jjtGetChild(i));
                 }
             }
         }
         
         if (null != o.myNode) {
-            
             JexlNode sourceNode = getSourceNode(o.myNode);
             JexlNode topLevelOr = getOrNode(sourceNode);
             
             if (null == topLevelOr) {
-                topLevelOr = sourceNode;
-                // add the source node
-                nodesMap.put(nodeToKey(topLevelOr), o.myNode);
+                nodeSet.add(o.myNode);
             } else {
                 for (int i = 0; i < topLevelOr.jjtGetNumChildren(); i++) {
-                    JexlNode baseNode = getSourceNode(topLevelOr.jjtGetChild(i));
-                    nodesMap.put(nodeToKey(baseNode), topLevelOr.jjtGetChild(i));
-                }
-            }
-        }
-        
-        for (JexlNode node : delayedNodes) {
-            JexlNode baseNode = getSourceNode(node);
-            nodesMap.put(nodeToKey(baseNode), node);
-        }
-        
-        for (String key : nodesMap.keySet()) {
-            Collection<JexlNode> nodeColl = nodesMap.get(key);
-            JexlNode delayedNode = null;
-            if (nodeColl.size() > 1) {
-                for (JexlNode node : nodeColl) {
-                    if (isDelayed(node)) {
-                        delayedNode = node;
-                        break;
-                    }
+                    nodeSet.add(topLevelOr.jjtGetChild(i));
                 }
             }
-            if (null != delayedNode)
-                internalNodeList.add(delayedNode);
-            else
-                internalNodeList.add(nodeColl.iterator().next());
         }
         
-        // ensure that we already aren't in there
+        // Add delayed nodes.
+        nodeSet.addAll(delayedNodes);
         
         merged.count = -1;
         merged.uids = ImmutableSortedSet.of();
-        if (internalNodeList.isEmpty()) {
-            
+        if (nodeSet.isEmpty()) {
             merged.myNode = null;
         } else {
-            
-            merged.myNode = JexlNodeFactory.createUnwrappedOrNode(internalNodeList);
+            merged.myNode = JexlNodeFactory.createUnwrappedOrNode(nodeSet.getNodes());
```

Comment:
This call will reparent the nodes returned by nodeSet.getNodes().  Are we sure we want to do that, or should we create copies of these nodes, so that we don't change the lineage of the original tree?

Reply:
This preserves the original implementation, but I like the idea of copying the nodes. I can experiment with that.